design/func-version.md

```diff
@@ -0,0 +1,378 @@
+# Kanister Functions Version Management
+
+<!-- toc -->
+- [Problem Statement](#problem-statement)
+- [Proposed Solution](#proposed-solution)
+  - [Summary](#summary)
+  - [Assumptions And Constraints](#assumptions-and-constraints)
+  - [Function Versioning](#function-versioning)
+  - [Function Registration And Discovery](#function-registration-and-discovery)
+  - [User Experience](#user-experience)
+  - [Version Promotion And Deprecation](#version-promotion-and-deprecation)
+  - [Versions Discovery](#versions-discovery)
+- [Test Cases](#test-cases)
+<!-- /toc -->
+
+## Problem Statement
+
+As Kanister continues to grow and evolve to meet the needs of its community,
+it's inevitable that eventually, some of the existing built-in [Kanister
+Functions][0] will be updated with breaking changes, deprecated or even
+removed entirely.
+
+The lack of a proper strategy to manage and rollout such changes in a
+responsible backwards compatible manner can lead to costly operational overhead
+for Kanister users. In addition, it also becomes difficult for Kanister
+maintainers to assess the blast radius of the change, determine the rollout
+timing, provide timely support etc.
+
+This proposal outlines a version management scheme that can be used to manage
+the lifecycle of Kanister Functions as they evolve from inception to retirement.
+
+## Proposed Solution
+
+### Summary
+
+The proposed solution uses a version management scheme to communicate the
+stability, availability, promotion and deprecation of Kanister Functions.
+
+The design extends the existing implementation of the action-level
```

Comment:
Function-level settings take precedence.

Reply:
I somehow think that we should try to simplify this, if not as part of this PR then maybe as part of other PR. Because there seem to be a lot of options now using which we can specify the function version which might become confusing to users as well as developers. Or do you think it's ok?

Reply:
Removing field from the API requires a bit more planning. Personally, I want to remove the actionset level option but I think it's beyond the scope of this design.